### ISSUE TITLE
MMR over time graph

### ISSUE THREAD

**Issue by Cepheid-UK** [w3champions/w3champions-ui#215]
A great feature for profiles would be some player-specific stats.

One in particular that has been requested multiple times is an MMR-over-time graph for the season.

Example:
![mmr-over-time](https://user-images.githubusercontent.com/40397054/96904418-ff88d080-148e-11eb-868d-4c37a3197fbf.png)

**Comment by weidenbach** [comment]
![grafik](https://user-images.githubusercontent.com/73471359/99110594-eafdab00-25ea-11eb-888e-f5ed29f3c099.png)
I'm making good progress. Will request PR soon.

**Comment by weidenbach** [comment]
![grafik](https://user-images.githubusercontent.com/73471359/99110875-5a739a80-25eb-11eb-9486-9f01b8d2eab5.png)
My current intend is to place it beneath the race winrate grid in the statistics tab of the player profile.

**Comment by Cepheid-UK** [comment]
If it's not too much, a few requests:
1. Can we make it a line without the area under it, the area itself doesn't have any real meaning.
2. If possible can we allow the user to switch between RP and MMR?
3. The tooltip when you hover a point on the graph should ideally show the percentile of the player at that time. Although this may not be trivial/worth as I can't remember if we store historical quantile data or not.

**Comment by weidenbach** [comment]
Hey, thanks!
>     1. Can we make it a line without the area under it, the area itself doesn't have any real meaning.
Sure, you have a point. I copied the styling from the Players per day graph on the statistics page. 
>     2. If possible can we allow the user to switch the graph between RP and MMR via a checkbox? (I'd rather not display both together personally)
For this the backend needs to be adapted, as it only supports MMR as of now. But this shouldn't be a problem.
I also thought about adding RP and thought first I'd work on the MMR to see if everything works. I was also considering showing both MMR and RP. I could do it like "Games per Day" in the statistics page, where you can toggle the curves of the graph and set RP to be disabled initially. This allows to view both at the same time or each individually.
>     3. The tooltip when you hover a point on the graph should ideally show the percentile of the player at that time. Although this may not be trivial/worth as I can't remember if we store historical quantile data or not.
Hmm, maybe its a bit unwieldy if its only in the tooltip. Perhaps also as a seperate curve, disabled by default? The backend doesn't support this I believe, but I can look into how feasible it would be. Hope that doesn't take too long. Might be way more effort than its worth.

**Comment by alocate** [comment]
hey, thanks a lot, it's really a feature a lot of people want! 

Few improvements from my side:
-how many data points are there for a given day? I think it'd benefit from considering only one (well chosen) data point per day.
-it should be zoomed in one something like [min-100 mmr, max+100 mmr] 

For the future:
-how to deal with resets? Would be nice to be able to pull the reset dates and draw a dashed vertical line where they happened.

**Comment by weidenbach** [comment]
Hey alocate!

> -how many data points are there for a given day? I think it'd benefit from considering only one (well chosen) data point per day.
Currently every change of mmr is a point. The line graph above is from ToD, which is why there are so many, as he plays a lot. So for most players it will not look as crowded. If we were to use only one point per day we'd lose fidelity, so it's definitely a tradeoff. Not sure whats best. It seems like for players like ToD it would definitely be too much data to display. The graph is covering merely a month so for a whole season it would be even more drastic. It should be limited in that case. One data point per day is certainly an option, but what if players only play during weekends? The data could be pretty scarce for some cases. I'd go for a threshold for points displayed and evenly distribute the data. Then it's not limiting when it doesn't need to.
> -it should be zoomed in on something like [min mmr-100, max mmr+100], this will greatly improve readability!
Definitely, good point. 
> For the future:
> -how to deal with resets? Would be nice to be able to pull the reset dates and draw a dashed vertical line where they happened.
Currently I only planend to offer the graph for a certain season, but an overarching graph would also be nice I guess. So I didn't even consider it. Then it would be nice to highlight each season with a unique background color. This would also clearly show what what caused the dip in MMR, as MMR resets happen at the start of a season.

**Comment by Cepheid-UK** [comment]
There is an argument to say it would be nice to have season-specific graphs that you can check, so I'd be fine with a dropdown that lets you pick which season to look at, or we could just make it a rolling continuous graph but with a marker to indicate when the season was reset.

**Comment by weidenbach** [comment]
![grafik](https://user-images.githubusercontent.com/73471359/99309375-9b70e680-2859-11eb-81b1-0ab42d380f2b.png)
I played a bit with the design. The graph is a lot more clear, because the y axis doesn't start at 0 anymore.
I also added a gradient effect instead of the plain background. While no background would be the most readible, this adds some extra eyecandy and it looks more interesting imo. I also resized the points a bit to clutter it less.

**Comment by abeninski** [comment]
Can we release it already. Looks super good. :P

**Comment by modmoto** [comment]
are we missing something still? There was some stuff merged, but i think the ui is missing the pr still?

**Comment by weidenbach** [comment]
I'm still working on the functunality, as there are some things not done. Some of which would be better to have before the PR I think. 

-  Set most played race to default.
-  The graph dates are messy, because it doesn't take the time between dates into account. So breaks aren't visible.
-  Season overarching timeline is missing
- RP isn't displayed at all

I'll try to get the first two points done today and create a PR.

**Comment by modmoto** [comment]
ok nice

**Comment by weidenbach** [comment]
Regarding:
- The graph dates are messy, because it doesn't take the time between dates into account. So breaks aren't visible.

![grafik](https://user-images.githubusercontent.com/73471359/99533103-a20c7480-29a5-11eb-89f1-a58b6bddbb8c.png)
This just looks too messy so I will not keep this

**Comment by alocate** [comment]
agreed - the other version looks a lot better to me. it looks super nice though!

**Comment by modmoto** [comment]
maybe an aggregated mmr per day? Just make a cut at 0:00 as one point

**Comment by weidenbach** [comment]
#

> 
> 
> maybe an aggregated mmr per day? Just make a cut at 0:00 as one point. dont need it for every game, i guess. Also if you play a lot, you will see only the progress within one day (when you play some 30 games on one day).

I'm open for changing this. I created a PR so we can release the graph to test. You can check the graph and how it appears on different player pages to see if displaying an aggregated mmr per day is needed / beneficial. That change definitely has pros and cons.

**Comment by Cepheid-UK** [comment]
chartjs has some nice options for smoothing the point lines (e.g. the activity graph we have in the statistics) that might help it look a bit less severe.

I can add it to test.w3champions.com, just @ me on the PR when (if already) its ready for test.

**Comment by modmoto** [comment]
i like the rounded edges, tbh. But it should just stay consistent.

**Comment by weidenbach** [comment]
I'm aware of the line smoothing, the default setting actually has line smoothing enabled. I set it to 0 with "lineTension: 0.0,", because I wanted to have this sharp effect. It shows the ups and downs better and helped declutter a bit, which was needed for some of the more dense graphs. But I can see that it breaks the consistency so I don't mind changing it.
@Cepheid-UK I added a PR yesterday and that should be good to go for test. Just note that it won't work as of now because the MmrTimelineHandler is disabled in the backend.
